src/transformers/models/rt_detr/modeling_rt_detr.py

```diff
@@ -0,0 +1,1652 @@
+        for feat in projected_features:
+            height, width = feat.shape[-2:]
+            # [batch, channels, height, width] -> [batch, height*width, channel]
+            feat_flatten.append(feat.flatten(2).permute(0, 2, 1))
+            # [num_levels, 2]
+            spatial_shapes.append([height, width])
+            # [level], start index of each level
+            level_start_index.append(height * width + level_start_index[-1])
+
+        # [batch, level, channel]
+        feat_flatten = torch.concat(feat_flatten, 1)
+        level_start_index.pop()
+
+        # prepare denoising training
+        if self.training and self.num_denoising > 0 and targets is not None:
+            denoising_class, denoising_bbox_unact, attn_mask, dn_meta = get_contrastive_denoising_training_group(
+                targets,
+                self.num_classes,
+                self.num_queries,
+                self.denoising_class_embed,
+                num_denoising=self.num_denoising,
+                label_noise_ratio=self.label_noise_ratio,
+                box_noise_scale=self.box_noise_scale,
+            )
+        else:
+            denoising_class, denoising_bbox_unact, attn_mask, dn_meta = None, None, None, None
+
+        batch_size = len(feat_flatten)
+        device = feat_flatten.device
+
+        # prepare input for decoder
+        if self.training or self.eval_spatial_size is None:
+            anchors, valid_mask = self.generate_anchors(spatial_shapes, device=device)
+        else:
+            anchors, valid_mask = self.anchors.to(device), self.valid_mask.to(device)
+
+        memory = valid_mask.to(feat_flatten.dtype) * feat_flatten
+
+        output_memory = self.enc_output(memory)
+
+        enc_outputs_class = self.enc_score_head(output_memory)
+        enc_outputs_coord_unact = self.enc_bbox_head(output_memory) + anchors
+
+        _, topk_ind = torch.topk(enc_outputs_class.max(-1).values, self.num_queries, dim=1)
+
+        reference_points_unact = enc_outputs_coord_unact.gather(
+            dim=1, index=topk_ind.unsqueeze(-1).repeat(1, 1, enc_outputs_coord_unact.shape[-1])
+        )
+
+        enc_topk_bboxes = F.sigmoid(reference_points_unact)
+        if denoising_bbox_unact is not None:
+            reference_points_unact = torch.concat([denoising_bbox_unact, reference_points_unact], 1)
+
+        enc_topk_logits = enc_outputs_class.gather(
+            dim=1, index=topk_ind.unsqueeze(-1).repeat(1, 1, enc_outputs_class.shape[-1])
+        )
+
+        # extract region features
+        if self.learnt_init_query:
+            target = self.tgt_embed.weight.unsqueeze(0).tile([batch_size, 1, 1])
+        else:
+            target = output_memory.gather(dim=1, index=topk_ind.unsqueeze(-1).repeat(1, 1, output_memory.shape[-1]))
+
+        if denoising_class is not None:
+            target = torch.concat([denoising_class, target], 1)
+
+        init_ref_points_unact = reference_points_unact.detach()
```

Comment:
yes. Although it makes no difference using the `reference_points_unact` and its detached version during inference. I'm not sure if that affects the training. 
Should I use directly the `reference_points_unact` instead of its detached version?